Avoid sorting retrieved font information

pvollan opened the pull request:
#### 776a973002faf35d63116e903eeb8a10aefae783
<pre>
Avoid sorting retrieved font information
<a href="https://bugs.webkit.org/show_bug.cgi?id=296856">https://bugs.webkit.org/show_bug.cgi?id=296856</a>
<a href="https://rdar.apple.com/157328710">rdar://157328710</a>

Reviewed by Sihui Liu.

This patch improves performance by adding the option to not sort the retrieved font data,
since sorting is not required in our use case.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:

Canonical link: <a href="https://commits.webkit.org/298251@main">https://commits.webkit.org/298251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5cadfe7ed1bbc2bce08927bab8a51420f2d38f2e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114830 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25039 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120992 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65520 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/706f9f36-1cef-4275-b81b-64f33fdda4f7) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43133 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87278 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/eb71d31e-6d01-4234-b148-2681d643a3f0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117778 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28052 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103116 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67670 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27237 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21240 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/64656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/107177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97437 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124191 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113386 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41834 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31259 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42212 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99305 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/95876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41053 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/18896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18381 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41715 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47229 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/137600 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41276 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/36790 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44590 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43020 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

basuke reviewed (commented):
How are these singleton resources protected for threads? Is this code run only in main thread?

aprotyas commented on Source/WTF/wtf/spi/cocoa/XTSPI.h:
```diff
@@ -36,3 +36,16 @@ enum {
+    kXTOptionsIncludeDisabledFonts      = 1 << 5,
+    kXTOptionsDisableFontProviders      = 1 << 16,
+    kXTOptionsDisableAutoActivation     = 1 << 17
+};
```
I think you can drop the unused options

```suggestion
enum {
    kXTOptionsDoNotSortResults          = 1 << 4
};
```

pvollan replied:
That is a good point, fixed in latest patch.

Thanks for reviewing!

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1533,18 +1553,18 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
 {
-    if (m_userInstalledFontURLs) {
-        process.send(Messages::WebProcess::RegisterFontMap(*m_userInstalledFontURLs, *m_userInstalledFontFamilyMap,  sandboxExtensionsForFonts(*m_sandboxExtensionURLs, process.auditToken())), 0);
+    if (userInstalledFontURLs()) {
```
We should capture it in local variable:
`if (auto fontURLs = userInstalledFontURLs()) { }`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -240,6 +240,26 @@ static void registerUserDefaults()
     return cachedLockdownModeEnabledGlobally;
 }
 
+static std::optional<HashMap<String, URL>>& userInstalledFontURLs()
```
I think we might want to add "shared" in the names to make it clear these are singleton.

szewai reviewed (commented):
`_registerAdditionalFonts` is an SPI on `WKProcessPool` so it seems client can register different additional fonts for different process pools with current implementation. Have we confirmed the existing usage will not break if the urls are shared across process pools?

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1510,18 +1536,18 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
     Vector<SandboxExtension::Handle> handles;
-    if (m_userInstalledFontURLs)
-        handles.appendVector(sandboxExtensionsForFonts(m_userInstalledFontURLs->values(), std::nullopt));
+    if (sharedUserInstalledFontURLs())
```
`if (auto& fontURLs = sharedUserInstalledFontURLs()) {`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -1510,18 +1536,18 @@ static void webProcessPoolHighDynamicRangeDidChangeCallback(CFNotificationCenter
 
-    if (m_sandboxExtensionURLs)
-        handles.appendVector(sandboxExtensionsForFonts(*m_sandboxExtensionURLs, std::nullopt));
+    if (sharedSandboxExtensionURLs())
```
`if (auto& sandboxURLs = sharedSandboxExtensionURLs()) {`

szewai commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -240,6 +240,32 @@ static void registerUserDefaults()
+    return lock;
+}
+
+static std::optional<HashMap<String, URL>>& sharedUserInstalledFontURLs() WTF_REQUIRES_LOCK(userInstalledFontDataLock())
```
It seems we didn't have the lock before, why do we need it now?

szewai reviewed (approved):
LGTM with nits

pvollan committed:
Avoid sorting retrieved font information
https://bugs.webkit.org/show_bug.cgi?id=296856
rdar://157328710

Reviewed by Sihui Liu.

This patch improves performance by adding the option to not sort the retrieved font data,
since sorting is not required in our use case.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:

Canonical link: https://commits.webkit.org/298251@main
